exporter/metadataexporter/cache.go

```diff
@@ -0,0 +1,283 @@
+		ttlcache.WithCapacity[string, struct{}](uint64(opts.LogsFingerprintCacheSize)),
+	)
+	go logsFingerprintCache.Start()
+
+	return &InMemoryKeyCache{
+		tracesFingerprintCache:  tracesFingerprintCache,
+		metricsFingerprintCache: metricsFingerprintCache,
+		logsFingerprintCache:    logsFingerprintCache,
+		tenantID:                opts.TenantID,
+		logger:                  opts.Logger,
+	}
+}
+
+func getKeyForInMemory(tenantID string, ds pipeline.Signal, key string) string {
+	var newKey strings.Builder
+	newKey.WriteString(getKey(tenantID, ds))
+	newKey.WriteString(":")
+	newKey.WriteString(key)
+	return newKey.String()
+}
+
+func (c *InMemoryKeyCache) Add(ctx context.Context, key string, ds pipeline.Signal) error {
+	key = getKeyForInMemory(c.tenantID, ds, key)
+	switch ds {
+	case pipeline.SignalTraces:
+		c.tracesFingerprintCache.Set(key, struct{}{}, ttlcache.DefaultTTL)
+	case pipeline.SignalMetrics:
+		c.metricsFingerprintCache.Set(key, struct{}{}, ttlcache.DefaultTTL)
+	case pipeline.SignalLogs:
+		c.logsFingerprintCache.Set(key, struct{}{}, ttlcache.DefaultTTL)
+	}
+	return nil
+}
+
+func (c *InMemoryKeyCache) Exists(ctx context.Context, key string, ds pipeline.Signal) (bool, error) {
+	key = getKeyForInMemory(c.tenantID, ds, key)
+	switch ds {
+	case pipeline.SignalTraces:
+		return c.tracesFingerprintCache.Get(key) != nil, nil
+	case pipeline.SignalMetrics:
+		return c.metricsFingerprintCache.Get(key) != nil, nil
+	case pipeline.SignalLogs:
+		return c.logsFingerprintCache.Get(key) != nil, nil
+	}
+	return false, nil
+}
+
+func (c *InMemoryKeyCache) ExistsMulti(ctx context.Context, keys []string, ds pipeline.Signal) ([]bool, error) {
+	exists := make([]bool, len(keys))
+	for i, key := range keys {
+		exists[i], _ = c.Exists(ctx, key, ds)
+	}
+	return exists, nil
+}
+
+func (c *InMemoryKeyCache) AddMulti(ctx context.Context, keys []string, ds pipeline.Signal) error {
+	for _, key := range keys {
+		c.Add(ctx, key, ds)
+	}
+	return nil
+}
+
+func (c *InMemoryKeyCache) Debug(ctx context.Context) {
+	c.logger.Debug("IN MEMORY KEY CACHE DEBUG")
+	c.logger.Debug("TRACES", zap.Strings("keys", c.tracesFingerprintCache.Keys()))
+	c.logger.Debug("METRICS", zap.Strings("keys", c.metricsFingerprintCache.Keys()))
+	c.logger.Debug("LOGS", zap.Strings("keys", c.logsFingerprintCache.Keys()))
+}
+
+type RedisKeyCache struct {
+	redisClient *redis.Client
+	tracesTTL   time.Duration
+	metricsTTL  time.Duration
+	logsTTL     time.Duration
+
+	tenantID string
+
+	logger *zap.Logger
+}
+
+type RedisKeyCacheOptions struct {
+	Addr                       string
+	Username                   string
+	Password                   string
+	DB                         int
+	TracesFingerprintCacheTTL  time.Duration
+	MetricsFingerprintCacheTTL time.Duration
+	LogsFingerprintCacheTTL    time.Duration
+	TenantID                   string
+	Logger                     *zap.Logger
+}
+
+func NewRedisKeyCache(opts RedisKeyCacheOptions) *RedisKeyCache {
```

Comment:
```suggestion
func NewRedisKeyCache(opts RedisKeyCacheOptions) (*RedisKeyCache, error) {
....
	if err := redisKeyCache.Ping(context.Background()).Err(); err != nil {
		return nil, err
	}
....
```

When we call `redis.NewClient`, it only initialises the struct and doesn't check whether the input details correspond to a valid redis connection or not. We should add a `Ping` to confirm.

Reply:
no error return policy in this exporter, but since this is the init stage, I will add it.